qt/python/mantidqt/project/recovery/projectrecovery.py

```diff
@@ -0,0 +1,458 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+#  This file is part of the mantidqt package
+#
+
+import os
+import getpass
+from threading import Timer
+import time
+import datetime
+from glob import glob
+import shutil
+import psutil
+
+from qtpy.QtCore import QMetaObject, Q_ARG, Qt
+from qtpy.QtWidgets import QApplication
+
+from mantid.kernel import ConfigService, logger, UsageService
+from mantid.api import AnalysisDataService as ADS, WorkspaceGroup
+from mantidqt.project.projectsaver import ProjectSaver
+from mantidqt.project.projectloader import ProjectLoader
+from mantidqt.project.recovery.recoverygui.projectrecoverypresenter import ProjectRecoveryPresenter
+from mantid.simpleapi import AlgorithmManager
+
+
+SAVING_TIME_KEY = "projectRecovery.secondsBetween"
+NO_OF_CHECKPOINTS_KEY = "projectRecovery.numberOfCheckpoints"
+RECOVERY_ENABLED_KEY = "projectRecovery.enabled"
+
+
+class ProjectRecovery(object):
+    def __init__(self, globalfiguremanager, window_finder, multifileinterpreter):
+        self.recovery_directory = os.path.join(ConfigService.getAppDataDirectory(), "workbench-recovery")
+        self.recovery_directory_hostname = os.path.join(self.recovery_directory, getpass.getuser())
+        self.recovery_directory_pid = os.path.join(self.recovery_directory_hostname, str(os.getpid()))
+
+        self.recovery_order_workspace_history_file = os.path.join(ConfigService.getAppDataDirectory(),
+                                                                  "ordered_recovery.py")
+
+        self.recovery_enabled = ConfigService[RECOVERY_ENABLED_KEY]
+        self.maximum_num_checkpoints = ConfigService[NO_OF_CHECKPOINTS_KEY]
+        self.time_between_saves = int(ConfigService[SAVING_TIME_KEY])  # seconds
+
+        self._timer_thread = Timer(self.time_between_saves, self.recovery_save)
+
+        self.recovery_file_ext = ".recfile"
+        self.lock_file_name = "projectrecovery.lock"
+
+        self.gfm = globalfiguremanager
+        self.interface_finding_func = window_finder
+        self.multi_file_interpreter = multifileinterpreter
+
+        # To ignore an algorithm in project recovery please put it's name here. e.g. MonitorLiveData is ignored because
```

Comment:
I add an example as to why later in the comment. For example on a recovery script you wouldn't want a save algorithm happening again etc etc etc.

Reply:
I can see the `MonitorLiveData` explanation, but it doesn't mean anything to me as I'm not familiar with that alg. 

> For example on a recovery script you wouldn't want a save algorithm happening again
I think this is a better comment to have above `self.algs_to_ignore`, i.e. ignoring an alg prevents undesirable calls to those algorithms, to prevent errors like repeated saving during recovery

Reply:
So I would need to add an explanation into the code for literally every type of algorithm that needs to be ignored?

Reply:
Say in the future we need to ignore CreateSampleWorkspace and Rebin2D for whatever reason... do they each need a detailed example for each... it seems a little overkill.

Reply:
Not for each, but to describe the overarching reason for ignoring an algorithm: prevent their execution when recovering

Reply:
I see!